File: pulsar/java/src/main/java/com/examples/hello/pulsar/producers/ProducerRunner.java
package com.examples.hello.pulsar.producers;

import com.examples.hello.pulsar.services.PulsarRunner;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.apache.pulsar.client.api.Producer;

@Slf4j
@RequiredArgsConstructor
public class ProducerRunner<T> extends PulsarRunner implements ProducerRunnable {
  private final Producer<T> producer;
  private final MessageProducer<T> source;

  @SneakyThrows
  @Override
  protected void callOnce() {
    val value = source.call();
    val id = producer.send(value);
    log.debug("Sent value={}, id={})", value, id);
  }

  @Override
  public void close() throws Exception {
    producer.close();
  }
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/producers/ProducerRunner.kt
package com.example.hello.pulsar.producers

import com.example.hello.pulsar.services.PulsarRunner
import org.apache.pulsar.client.api.Producer
import org.slf4j.LoggerFactory

class ProducerRunner<T>(
  private val producer: Producer<T>,
  private val source: () -> T
) : PulsarRunner(), ProducerRunnable {

    override fun callOnce() {
        val value = source.invoke()
        val id = producer.send(value)
        log.debug("Produced value={}, id={})", value, id)
    }

    @Throws(Exception::class)
    override fun close() {
        producer.close()
    }

    companion object {
        private val log = LoggerFactory.getLogger(ProducerRunner::class.java)
    }
}
